src/lightning_app/components/multi_node/base.py

```diff
@@ -52,31 +52,26 @@ def run(
             work_kwargs: Keywords arguments to be provided to the work on instantiation.
         """
         super().__init__()
-        self.ws = structures.List()
-        self._work_cls = work_cls
+        self.ws = structures.List(
+            *[
+                work_cls(
+                    *work_args,
+                    cloud_compute=cloud_compute,
+                    **work_kwargs,
+                    parallel=True,
+                )
+                for _ in range(num_nodes)
+            ]
+        )
         self.num_nodes = num_nodes
-        self._cloud_compute = cloud_compute
-        self._work_args = work_args
-        self._work_kwargs = work_kwargs
         self.has_started = False
 
     def run(self) -> None:
         if not self.has_started:
 
-            # 1. Create & start the works
-            if not self.ws:
-                for node_rank in range(self.num_nodes):
-                    self.ws.append(
-                        self._work_cls(
-                            *self._work_args,
-                            cloud_compute=self._cloud_compute,
-                            **self._work_kwargs,
-                            parallel=True,
-                        )
-                    )
-
-                    # Starting node `node_rank`` ...
-                    self.ws[-1].start()
+            # 1. Start all the works.
+            for w in self.ws:
+                w.start()
```

Comment:
Might not be needed since `start_with_flow=True` by default now.

Reply:
Great catch, but it is still needed, as we don't send the ip_address before the first execution. I was actually thinking to move the delta with the ip_address to the WorkRunner setup call.

So this could be removed and converted to:

```python
    def run(self) -> None:
        if not all(w.internal_ip for w in self.ws):
            return

        # Loop over all node machines
        for node_rank in range(self.num_nodes):

            # 3. Run the user code in a distributed way !
            self.ws[node_rank].run(
                main_address=self.ws[0].internal_ip,
                main_port=self.ws[0].port,
                num_nodes=self.num_nodes,
                node_rank=node_rank,
            )

            # 4. Stop the machine when finished.
            if self.ws[node_rank].has_succeeded:
                self.ws[node_rank].stop()
```

Reply:
Going to add the change to this PR.

Reply:
Done !